### PR TITLE
global: dump the siginfo when handling a fatal signal.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/global/signal_handler.cc [hunk 1]
```diff
@@ -173,6 +173,9 @@ static void handle_fatal_signal(int signum)
   BackTrace bt(1);
   ostringstream oss;
   bt.print(oss);
+  if (signum == SIGSEGV) {
+    oss << std::endl << "Fault at location: " << siginfo->si_addr << std::endl;
```
Comment (batrick) [inline src/global/signal_handler.cc:177]: While you're here, you might as well dump the whole struct. It's all useful information.

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by jdurgin** [comment]
@yaarith some of these extra fields from struct sig_info might be interesting for crash reports

**Comment by yaarith** [comment]
@jdurgin thanks! yes, we should take si_signo and si_code

File: src/global/signal_handler.h [hunk 2]
```diff
@@ -28,7 +29,7 @@ typedef void (*signal_handler_t)(int);
 # define sig_str(signum) sys_siglist[signum]
 #endif
 
-void install_sighandler(int signum, signal_handler_t handler, int flags);
+void install_sighandler(int signum, signal_action_handler_t handler, int flags);
```
Comment (tchaikov) [inline src/global/signal_handler.h:32]: this change breaks the win32 build.
```
../src/global/signal_handler.h:22:46: error: 'siginfo_t' has not been declared
   22 | typedef void (*signal_action_handler_t)(int, siginfo_t*, void*);
      |                                              ^~~~~~~~~
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=windows,DIST=windows,MACHINE_SIZE=gigantic/54417//consoleFull

Comment (rzarzynski) [reply src/global/signal_handler.h:32]: Oops, will add a guard. Thanks for pointing this out, Kefu!

**Comment by rzarzynski** [comment]
Sent the updated-for-`_WIN32` revision to shaman some time ago. Still waiting (https://shaman.ceph.com/builds/ceph/wip-global-backtrace-si_addr/).

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-12_04:48:43-rados-wip-kefu-testing2-2021-05-12-1103-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-13_03:34:40-rados-wip-kefu-testing-2021-05-12-2313-distro-basic-smithi/

**Comment by tchaikov** [comment]
@rzarzynski the win32 build still fails. but i cannot find anything suspicious after skimming through it. could you take another look into it?

**Comment by tchaikov** [comment]
@rzarzynski ping?

**Comment by badone** [comment]
Created https://tracker.ceph.com/issues/51435 to try to get this into nautilus and pacific for downstream consumption.

**Review by badone — CHANGES_REQUESTED** [review]
Looks good if we can resolve the WIN32 issue @rzarzynski ?

File: src/global/signal_handler.h [hunk 3]
```diff
@@ -28,7 +31,11 @@ typedef void (*signal_handler_t)(int);
 # define sig_str(signum) sys_siglist[signum]
 #endif
 
-void install_sighandler(int signum, signal_handler_t handler, int flags);
+#ifndef _WIN32
+void install_sighandler(int signum, signal_action_handler_t handler, int flags);
+#else
+void install_sighandler(int signum, signal_handler_t handler, int flags)
```
Comment (rzarzynski) [inline src/global/signal_handler.h:37]: Missed `;` on win32.

**Comment by badone** [comment]
Let's revisit this, since I think it's worth having? @tchaikov are you happy with it once it is rebased?

**Pull request closed without merging**